overseer/src/lib.rs

```diff
@@ -0,0 +1,552 @@
+/// The `Overseer` itself.
+pub struct Overseer<M: Debug> {
+	/// All `Subsystem`s by their respective `SubsystemId`s.
+	subsystems: HashMap<SubsystemId, OverseedSubsystem<M>>,
+	/// The actual poor man's process tree.
+	///
+	/// Needed (among other things) to stop a running `Job` along
+	/// with all it's children.
+	id_to_children: HashMap<SubsystemId, HashSet<SubsystemId>>,
+}
+
+impl<M: Debug + Clone> Overseer<M> {
+	/// Create a new intance of the `Overseer` with some initial set of `Subsystems.
+	///
+	/// The `Subsystems` submitted to this call will act as a level 1 in the "process tree":
+	///
+	///
+	/// ```text
+	///                  +------------------------------------+
+	///                  |            Overseer                |
+	///                  +------------------------------------+
+	///                    /            |             |      \
+	///      ................. subsystems[..] ..............................
+	///      . +-----------+    +-----------+   +----------+   +---------+ .
+	///      . |           |    |           |   |          |   |         | .
+	///      . +-----------+    +-----------+   +----------+   +---------+ .
+	///      ...............................................................
+	///                              |
+	///                        probably `spawn`
+	///                        something else
+	///                              |
+	///                              V
+	///                         +-----------+
+	///                         |           |
+	///                         +-----------+
+	///
+	/// ```
+	pub fn new(subsystems: Vec<Box<dyn Subsystem<M>>>) -> Self {
+		let mut this = Self {
+			subsystems: HashMap::new(),
+			id_to_children: HashMap::new(),
+		};
+
+		for s in subsystems.into_iter() {
+			let _ = this.spawn(s);
+		}
+
+		this
+	}
+
+	/// Run the `Overseer`.
+	// TODO: we have to
+	//   * Give out to the user some handler to communicate with the `Overseer`
+	//     to tell it to do things such as `Start` `Stop` or `Spawn`
+	//   * Actually implement stopping of the `Overseer`, atm it's unstoppable.
+	pub async fn run(mut self) {
+		loop {
+			// Upon iteration of the loop we will be collecting all the messages
+			// that need dispatching (if any).
+			let mut msgs = Vec::default();
+
+			for (id, s) in self.subsystems.iter_mut() {
+				if let Some(s) = &mut s.instance {
+					// TODO: set to `None` or restart.
+					// It will panic when the `Future` becomes
+					// ready and then it will be polled again here.
+					if let Poll::Ready(res) = poll!(&mut s.f.0) {
+						log::info!("subsystem stopped {:?}", res);
+					}
+
+					match poll!(&mut s.rx.next()) {
+						Poll::Ready(Some(msg)) => {
+							log::info!("Received message from subsystem {:?}", msg);
+							msgs.push((*id, msg));
+						}
+						_ => {}
+					}
+				}
+			}
+
+			// Do the message dispatching be it broadcasting or direct messages.
+			//
+			// TODO: this desperately need refactoring.
```

Comment:
it feels slightly misplaced to do all of this here, should not the `subsystem` take care of `processing` one `Message`

Reply:
The spec stresses out that the overseer itself does all the message routing to "provide a single source of heartbeat in the system" or I don't remember the exact words, but that piece of doc contains some motivation

Reply:
The heartbeat is about doing this rough thing

```rust
let new_chain_head = chain_events.next().await;
future::join_all(subsystems.iter().map(|s| s.send(StartWork(...)))).await;
// thus the `StartWork` message is buffered for all subsystems before we send any further message to subsystems
```